keps/0031-namespace-management.md

```diff
@@ -0,0 +1,167 @@
+
+### Hierarchical Namespace Controller
+
+Currently in incubation, Kubernetes is introducing a Hierarchical Namespace Controller to more easily create and manage namespaces in the cluster, even if you don't have cluster-level permission to create namespaces, which is something to keep in mind in the future.  See https://github.com/kubernetes-sigs/multi-tenancy/tree/master/incubator/hnc
+
+## Proposals
+
+Prior to making KEP `provisional` we will chose a namespace creation proposal to move forward with and move the 2nd proposal to alternative.  For now, we have 2 competing namespace creation proposals.  
+In addition there is a proposal for the use of multiple namespaces.
+
+### Proposal: Single Namespace Support
+
+There is a desire to support multiple namespaces, however it requires the addition of cluster-wide resource support in KUDO.  Based on this and the desire to not break Kubernetes garbase collection, KUDO currenly only supports 1 namespace.  Which results in the following explicit rules for operators until a cluster-wide support is provided:
+
+* All manifests with metadata defining a namespace will be ignored or overridden.  The namespace that is being used by KUDO to install the operator is the only namespace that resources will be installed to.
+* All namespace manifests (manifest of `"kind": "Namespace"`) will have no affect as they can not be used by the operator.
+* All namespace manifests will be flagged by `package verify` as a warning
+
+### Proposal: Namespace Creation
+
+KUDO should support the creation of a namespace if it does not exist.  Further KUDO will add a configuration for namespace such that at time of namespace creation it is created with the metadata provided. To this end a manifest file for namespace is to be supported in the template folder:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    ca.istio.io/override: "true"
+    istio-injection: enabled
+    katib-metricscollector-injection: enabled
+```
+
+The template supports templates for labels and annotations, however the `name:` metadata is NOT allowed to be set or controlled through templating.   This is to prevent the "templating" of the namespace name such as `name: {.Namespace}-extended`. Other templating of labels and metadata is allowed and is the reason the file lives under the template folder. KUDO will add the `name:` for metadata when there is a need to create a namespace with metadata using the namespaceManifest file. The `operator.yaml` file will be extended to support `namespaceManifest` attribute.  
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
```

Comment:
Do we want the `templates/` path in here? For all other resources in the `operator.yaml` we don't include this.

Reply:
I'd say we don't ^^

Reply:
I was expecting that an operator dev could use templates in the manifest file... which means it needs to be in the template folder.  there is no reason to prevent this that I can see... we just can't have it for the name itself

Reply:
Yes, the file should be in the templates folder an can use templating. It's only about the path:

If you look at the tasks:
```
tasks:
  - name: configmap
    kind: Apply
    spec:
      resources:
        - configmap.yaml
```

It doesn't say `templates/configmap.yaml` either.